### PR TITLE
Constrained tabbing: fix unstable behavior in firefox

### LINKED ISSUE #42652: Components: focus is lost from Modal on keyboard operation in Firefox
### Description

In the Firefox browser, the focus sometimes goes out of the modal f you use the TAB key to move.
This issue is very unstable and seems to occur when the machine is under load or when tabs are moved at high speed.

### Step-by-step reproduction instructions

- Open one of the modal window from [the Storybook page](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default) or "Welcome Guide" of the post editor, etc.
- Press `tab` key or `tab + shift` key to change the focus in the modal window.
- Alternatively, hold down `tab` key to change the focus to fast.
- Sometimes you should see the focus goes out of the modal.

## On Post Editor

https://user-images.githubusercontent.com/54422211/180588815-2936c9ab-6831-4d9c-8a8c-b03821a1fe37.mp4

## On Storybook

https://user-images.githubusercontent.com/54422211/180588813-18444700-e618-4909-a035-76dd0a660118.mp4

### Environment info

- WordPress Version: 6.1-alpha-53579
- Gutenberg Version: 13.7.2 (with or without)
- Browser Version: Firevox 102.0.1 (64bit)

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Fix #42652, #45903
Follow up on #34836

## What?
This PR fixes a problem with "constrained tabbing" sometimes losing focus in Firefox.

"Constrained tabbing" means that the tab moving is constrained to the content of the wrapper element such as modal component.

As reported in #42652, I confirmed that in Firefox, the focus is sometimes lost from the modal when the following operation is performed:

- When `tab` key is pressed on the **last** focusable element (the first element should be focused)
- When `tab + shift` key is pressed on the **first** focusable element (the last element should be focused)

## Why?

I learned that the constrained tabbing [useConstrainedTabbing](https://github.com/WordPress/gutenberg/blob/04802c3e1edef60a6df877d1a5006486564f65da/packages/compose/src/hooks/use-constrained-tabbing/index.js#L33-L76) hook.

For example, when the tab key is pressed on the last focused element, the following logic is used to move focus to the first element:

- Searches for the next element to focus from all tabbable elements
- Add a trap element to the wrapper element if the element is not contained in a wrapper element
- Focus on trap elements
- Browser default behavior moves focus to the next element after the trap element (i.e., the first element in the wrapper)
- Delete the trap element after the focus has successfully moved

However, in Firefox, I found that the trap element is sometimes deleted while the focus remains on the trap element.

```javascript
timeoutId = setTimeout( () => {
	// Sometimes the TRAP element still has focus.
	console.log( document.activeElement );
	node.removeChild( trap );
} );
```

In other words, I believe that the focus is lost by deleting the trap element that remains in focus, and the browser is moving focus without considering the wrapper element.

## How?
I haven't been able to find valid documentation, but I think we need to wait for the browser to properly move focus to the next element before removing the trap element.

So I added a small delay to the `setTimeout` function that removes the trap element.

## Testing Instructions

- Open a modal window such as "Welcome Guide" or "Preferences".
- Confirm that the focus isn't lost from within the modal when you press 'tab' or 'tab+shift' key.
- Or, hold down the tab key to move the focus fast.

## Screenshot

https://user-images.githubusercontent.com/54422211/180595778-5b89e7de-c376-4cae-bbbc-2cef3f4f1377.mp4

### CODE REVIEW

File: packages/compose/src/hooks/use-constrained-tabbing/index.js [hunk 1]
```diff
@@ -64,7 +64,7 @@ function useConstrainedTabbing() {
 			node[ domAction ]( trap );
 			trap.focus();
 			// Remove after the browser moves focus to the next element.
-			timeoutId = setTimeout( () => node.removeChild( trap ) );
+			timeoutId = setTimeout( () => node.removeChild( trap ), 10 );
```
Comment (ellatrix) [inline packages/compose/src/hooks/use-constrained-tabbing/index.js]: What if we add an event listener to the trap and remove the element only when focus was moved out of it?

Comment (t-hamano) [reply packages/compose/src/hooks/use-constrained-tabbing/index.js]: I have made the following changes and have confirmed that it works as expected:

```javascript
// Remove itself when the trap loses focus.
trap.addEventListener( 'blur', () => node.removeChild( trap ) );

trap.focus();
```

I couldn't figure out how to remove the event listener when the element itself is deleted, but is it also need to remove the event when the element is deleted?

File: packages/compose/src/hooks/use-constrained-tabbing/index.js [hunk 2]
```diff
@@ -62,9 +62,11 @@ function useConstrainedTabbing() {
 
 			trap.tabIndex = -1;
 			node[ domAction ]( trap );
+
+			// Remove itself when the trap loses focus.
+			trap.addEventListener( 'blur', () => node.removeChild( trap ) );
+
 			trap.focus();
-			// Remove after the browser moves focus to the next element.
-			timeoutId = setTimeout( () => node.removeChild( trap ) );
```
Comment (afercia) [inline packages/compose/src/hooks/use-constrained-tabbing/index.js:67]: Since this timeout is not used any longer, the variable declaration at line 36 and the clearTimeout at line 75 should be removed as well.

File: packages/compose/src/hooks/use-constrained-tabbing/index.js (hunk 1 above)
Comment (afercia) [reply packages/compose/src/hooks/use-constrained-tabbing/index.js]: I'm not an expert but I think the element itself _and_ the event listener are retained in memory because there's still a reference to it. To test: dump the trap element to the console after it's removed from the DOM, see that it's still in memory::
```
			trap.addEventListener( 'blur', () => {
				node.removeChild( trap );
				console.log( trap );
			} );
```

Since a new `trap` element and event listener will be created at each Tab key press on the last and first tabbable elements, I suspect this could lead to some memory leak. Again, I'm not an expert but I think that declaring the `trap` element with `let` and then assigning it to `null` after it's removed could be an option e.g. `trap = null;`.
Would appreciate feedback from someone with better knowledge than me.

**Comment by afercia** [comment]
Once this gets reviewed, it would be great to add some E2E tests with Playwright and run them on both Chrome and Firefox. Such an important usability and accessibility feature should never break and there should be tests in place to make sure it doesn't. 

Kind reminder that this is a regression present since WordPress 5.9 - Gutenberg 11.8, which is almost one year ago.
/Cc @ellatrix @youknowriad @alexstine

**Comment by alexstine** [comment]
Agreed. 👍

File: packages/compose/src/hooks/use-constrained-tabbing/index.js (hunk 1 above)
Comment (stokesman) [reply packages/compose/src/hooks/use-constrained-tabbing/index.js]: > I couldn't figure out how to remove the event listener when the element itself is deleted, but is it also need to remove the event when the element is deleted?

It's my understanding that the listener does not need to be removed as long as there is no persisting reference to the element. It looks like in this case it'd be fine because each time the `onKeyDown` handler creates `trap`, the only closure around it is the blur listener itself. However, to be sure about it you could use the [`once`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#once) option when adding the listener:

```js
// Remove itself when the trap loses focus.
trap.addEventListener( 'blur', () => trap.remove(), { once: true } );
```
That example also switches to `remove` instead of `removeChild` for saving a few bits. Since IE11 doesn't need supported that should be safe.

Also I couldn't quickly find an outside source to back up my statements but there is this bit from the the jsdoc of `useRefEffect`: https://github.com/WordPress/gutenberg/blob/a12ad38cc0c2692f9c025c3d63e0a394a1e29461/packages/compose/src/hooks/use-ref-effect/index.ts#L20-L23
Which agrees and indicates that this hook itself could drop the cleanup function here:
https://github.com/WordPress/gutenberg/blob/a12ad38cc0c2692f9c025c3d63e0a394a1e29461/packages/compose/src/hooks/use-constrained-tabbing/index.js#L73-L77

**Review by stokesman — APPROVED** [review]
I tested and verified I can reproduce the issue on trunk and I cannot on this branch.

Once Andrea’s [comment](https://github.com/WordPress/gutenberg/pull/42653/files#r1028102933) is addressed this LGTM.

I'll approve anyway because I'm confident no corners will be cut. Awesome work spotting this issue and figuring out the fix @t-hamano 🏅.

**Comment by afercia** [comment]
Note: I'm working on a separate PR to move the existing a11y tests (which do have some tests for constrained tabbing) to Playwright so that we can run them with Firefox too. I'll also try to add some new tests.

File: packages/compose/src/hooks/use-constrained-tabbing/index.js (hunk 2 above)
Comment (t-hamano) [reply packages/compose/src/hooks/use-constrained-tabbing/index.js:67]: Removed by 11a1cfa

**Comment by t-hamano** [comment]
Thank you for the review, @afercia, @stokesman, @alexstine!
I have removed the unnecessary variables and `clearTimeout` indicated in [this comment](https://github.com/WordPress/gutenberg/pull/42653#discussion_r1028102933).

@stokesman 
Should I add `{ once:true }` as suggested in [this comment](https://github.com/WordPress/gutenberg/pull/42653#discussion_r1028465993) just in case? As far as I have tested, it seems to work well in both cases.

**Comment by stokesman** [comment]
> Should I add `{ once:true }` as suggested in [this comment](https://github.com/WordPress/gutenberg/pull/42653#discussion_r1028465993) just in case? As far as I have tested, it seems to work well in both cases.

I don't think it’s needed. I largely bought it up to promote awareness of it in case it comes in handy elsewhere.

**Comment by youknowriad** [comment]
Thanks for the fix. Highly appreciated :)

**Comment by afercia** [comment]
🎉 ❤️
